File: Algorithms_python/108_Convert Sorted Array to Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> Optional[TreeNode]:
        def build(left, right):
            if left > right:
                return None
            p = (left + right) // 2
            
            root = TreeNode(nums[p])
            root.left = build(left, p-1)
            root.right = build(p+1, right)
            return root
        return build(0, len(nums)-1)
'''
选择左中间元素：p = (left + right) // 2
启动根：root = TreeNode(nums[p])
递归计算左右子树：root.left = helper(left, p - 1)
root.right = helper(p + 1, right)
'''

File: Algorithms_C++/108_Convert_Sorted_Array_to_Binary_Search_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& nums) {
        return dfs(0, nums.size()-1, nums);
    }
    TreeNode* dfs(int l, int r, vector<int>& nums){
        if(l > r){
            return nullptr;
        }
        int mid = (l + r) >> 1;
        TreeNode* left = dfs(l, mid - 1, nums);
        TreeNode* right = dfs(mid + 1, r, nums);
        return new TreeNode(nums[mid], left, right);
    }
};